CIP51: Federated Attestations Protocol (#314)
celo-org/celo-proposals: CIPs/cip-0051.md

Ensuring interoperability of attestations across the ecosystem is a key component of our vision. Currently, phone number identifier → address mappings (attestations) are stored in `Attestations.sol`. Anyone can query ODIS for the phone number identifier, in order to perform the on-chain lookup of stored attestations, but the reverse is not true: phone number attestations that are not created through integrating with ASv1 are not currently discoverable from other dApps. The proposed Federated Attestation Protocol aims to reconcile the benefits of dApps having full control over their verification flow with the benefits of on-chain interoperability. Alongside `Attestations.sol` we implement a new on-chain general attestation registry `FederatedAttestations.sol`, which will allow anyone to issue and publish their own identifier ↔ address mappings. While any account can write to the registry contract, wallets/dApps can choose to read only mappings from issuers they trust.

## Implementation
### Identifiers

> While issuers can choose to use identifiers in any format, we recommend using ODIS to derive identifiers that will be published on-chain. This offers a privacy-preserving standard that also allows for discoverability and interoperability. ODIS will be upgraded with the relase of ASv2 to include prefix types beyond phone numbers to accomodate all identifier types. 

Reviewer: Nit: typo "relase"

---
cip: 51
title: Federated Attestations Protocol
author: @eelanagaraj, @isabellewei
discussions-to: <URL>
status: Draft
type: Standards
category (*only required for Standards Track): Ring 0
created: 2022-06-29
license: Apache 2.0
---

## Overview

> This CIP proposes a new version (from here referred to as ASv2) of the existing validator-run attestation protocol (CIP01, from here referred to as ASv1), which will leverage a federated set of independent issuers to perform attestations. It will also extend attestations beyond phone numbers to other types of off-chain identifiers (email addresses, social media accounts, etc.)

Reviewer 2: ```suggestion
This CIP proposes a new version (from here referred to as ASv2) of the existing validator-run phone number attestation protocol (CIP01, from here referred to as ASv1), which will leverage a federated set of independent issuers to perform attestations. It will also lay the foundation for new identifier types (email addresses, social media accounts, etc.) to be brought on-chain.
```

Ensuring interoperability of attestations across the ecosystem is a key component of our vision. Currently, phone number identifier → address mappings (attestations) are stored in `Attestations.sol`. Anyone can query ODIS for the phone number identifier, in order to perform the on-chain lookup of stored attestations, but the reverse is not true: phone number attestations that are not created through integrating with ASv1 are not currently discoverable from other dApps. The proposed Federated Attestation Protocol aims to reconcile the benefits of dApps having full control over their verification flow with the benefits of on-chain interoperability. Alongside `Attestations.sol` we implement a new on-chain general attestation registry `FederatedAttestations.sol`, which will allow anyone to issue and publish their own identifier ↔ address mappings. While any account can write to the registry contract, wallets/dApps can choose to read only mappings from issuers they trust.

## Implementation

### Identifiers

> While issuers can choose to use identifiers in any format, we recommend using ODIS to derive identifiers that will be published on-chain. This offers a privacy-preserving standard that also allows for discoverability and interoperability. ODIS will be upgraded with the release of ASv2 to include prefix types beyond phone numbers to accomodate all identifier types.

Reviewer 2: ```suggestion
While issuers can choose to use identifiers in any format, we recommend using ODIS to derive identifiers that will be published on-chain. This offers a privacy-preserving standard that also allows for discoverability and interoperability. ODIS will be upgraded with the release of ASv2 to support new domains (see CIP40) for arbitrary identifier types (email address, twitter handles, etc.) 
```

In the future, we plan to allow for issuers to create attestations for the fact that a given identifier was derived using the “canonical” method described above, through the addition of metadata attestations. Identifiers derived using this “canonical” method could be used as a lightweight sybil resistant signal, as there is a one-to-one mapping between off-chain identities (such as phone numbers) and the derived identifier.

### Issuers

> Issuers should create a separate signer key that is associated with their account for the express purpose of signing verifiable credentials. The `AttestationSigner` role should be assigned to this signer key. It is possible to authorize multiple signer keys for this role.

Reviewer: Possible typo: 

> for the express purpose of signing verifiable credentials

Presumably to sign "attestations" instead of "verifiable credentials"?

## Further Work

We aim to later introduce metadata attestations in addition to ownership attestations. This would enable further information about an identifier to be claimed. Examples include phone number type, Twitter verification status, compliance with ODIS, etc. Furthermore, a more extensive system was originally imagined, where issuers issue [verifiable credentials](https://www.w3.org/TR/vc-data-model/) and identifiers follow the [DID standard](https://www.w3.org/2019/08/did-20190828/). These credentials are custodied by the user, and optionally can be shared on-chain to be discoverable by other parties, or can be shared off-chain directly between stakeholders.

To allow for users to easily decide which issuers to trust, there are plans for designing smart contracts and corresponding tooling to create [crowdsourced trusted issuer registries](https://www.notion.so/clabsco/Decentralisation-and-Governance-d673afd33dbe4f8d8b70daa2ef6d4763).

We hope that the information in these attestations can later be used to create lightweight reputation and sybil resistance signals for various use cases.

## Further Reading

> - [Federated Attestations Protocol](https://clabsco.notion.site/ASv2-On-chain-Registry-Protocol-47e7850656444c37973856e0634ee2f9) is the main high-level design doc with our intial overarching vision

Reviewer 3: @0xarthurxyz hey one note on this, this actually links to the protocol design doc; we should probably update the link to point to https://clabsco.notion.site/Federated-Attestation-Protocol-ASv2-05dc4843139842768ad9fe192cb46c00
and then add another entry to point to the protocol design doc (linked)

Reviewer: Thanks for catching that @eelanagaraj! 🕵️‍♀️

### Issuers

> Issuers should create a separate signer key that is associated with their account for the express purpose of signing attestations. The `AttestationSigner` role should be assigned to this signer key. It is possible to authorize multiple signer keys for this role.

Reviewer: @alecps: Made this tiny change to fix what looks like a typo. 
Let me know if you disagree 🙏

Reviewer: I agree 👍 That's why I replaced "verifiable credentials" with "attestations".

Author: oh woops yes true thank you haha
